streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/bolt/query/RealtimeJoinBolt.java

```diff
@@ -0,0 +1,638 @@
+            String key = makeLookupTupleKey(tuple);
+            if(dropOlderDuplicates)
+                lookupBuffer.removeAll(key);
+            lookupBuffer.put(key, new TupleInfo(tuple) );
+
+            if(timeBasedRetention) {
+                timeTracker.add(System.currentTimeMillis());
+            } else {  // count based Rotation
+                if (lookupBuffer.size() > retentionCount) {
+                    TupleInfo expired = removeHead(lookupBuffer);
+                    if( (joinType==JoinType.RIGHT) || (joinType==JoinType.OUTER)  ) {
+                        emitUnMatchedTuples(expired);
+                    }
+                    collector.ack(expired.tuple);
+                }
+            }
+        } else if (isDataStream(streamId) ) {
+            List<TupleInfo> matches = matchWithLookupStream(tuple);
+            if (matches==null || matches.isEmpty() ) {  // no match
+                if (joinType==JoinType.LEFT ||  joinType==JoinType.OUTER ) {
+                    ArrayList<Object> outputTuple = doProjection(tuple, null);
+                    emit(outputTuple, tuple);
+                    return;
+                }
+                collector.ack(tuple);
+            }
+
+            for (TupleInfo lookupTuple : matches) { // match found
+                lookupTuple.unmatched = false;
+                ArrayList<Object> outputTuple = doProjection(lookupTuple.tuple, tuple);
+                emit(outputTuple, tuple, lookupTuple.tuple);
+            }
+            collector.ack(tuple);
+        } else {
+            LOG.warn("Received tuple from unexpected stream/source : {}. Tuple will be dropped.", streamId);
+        }
+    }
+
+    private String makeLookupTupleKey(Tuple tuple) {
+        StringBuffer key = new StringBuffer();
+        for (JoinInfo ji : joinCriteria) {
+            String partialKey = findField(ji.lookupField, tuple).toString();
+            key.append( partialKey );
+            key.append(".");
+        }
+        return key.toString();
+    }
+
+    private String makeDataTupleKey(Tuple tuple) {
+        StringBuffer key = new StringBuffer();
+        for (JoinInfo ji : joinCriteria) {
+            String partialKey = findField(ji.dataField, tuple).toString();
+            key.append( partialKey );
+            key.append(".");
+        }
+        return key.toString();
+    }
+
+    // returns null if no match
+    private List<TupleInfo> matchWithLookupStream(Tuple lookupTuple) {
+        String key = makeDataTupleKey(lookupTuple);
+        return lookupBuffer.get(key);
+    }
+
+    // Removes timedout entries from lookupBuffer & timeTracker. Acks tuples being expired.
+    private void expireAndAckTimedOutEntries(LinkedListMultimap<String, TupleInfo> lookupBuffer) {
+        Long expirationTime = System.currentTimeMillis() - retentionTime;
+        Long  insertionTime = timeTracker.peek();
+        while ( insertionTime!=null && expirationTime.compareTo(insertionTime) > 0) {
```

Comment:
Can not we have `expirationTime > insertionTime` instead of using `compareTo` API? I may be missing something here.